### PR TITLE
CalorimeterHitDigi: unified code path for digitization with and without the hit merging

### PR DESCRIPTION
**Pull request by veprbl**
### Briefly, what does this PR introduce?
This should be easier to maintain. We probably could even refactor merging into
a separate algorithm, if we wanted to.

A side effect of this is that non-merging digitization now reorders hits (the
cell id is used as a key for an std::unordered_map).

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [ ] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [x] Changes have been communicated to collaborators @Chao1009 

### Does this PR introduce breaking changes? What changes might users need to make to their code?
No

### Does this PR change default behavior?
Yes, the hits will be reordered.

### CODE REVIEW

**Comment by veprbl** [comment]
This includes #856 to impose timing cut on hits that were previously routed through (now removed) non-merging path.

**Comment by Chao1009** [comment]
Refactoring a separate merging algorithm should be better, but we need to figure out how to pass the information from merging to digitization.

Current workflow:
`sim_hits (input) -> [merging -> digitization] -> digi_hits (output)`
Make merging a separate algorithm:
`sim_hits (input) > merging -> merged_info (intermediate hits) -> digitization -> digi_hits (output)`

"merged_info" could be the merged simulation hits, but this may rewrite a bunch of redundant information. It can also be some collection like the protocluster, which needs a modification to the data model.

**Comment by veprbl** [comment]
(That discussion beyond this PR) For the outputs of this envisioned factory that could take on hit energy deposition summation, can't use the same `edm4hep::RawCalorimeterHit` as for input?

**Comment by Chao1009** [comment]
Yes, the discussion is not for this PR : )
For the question regarding the `edm4hep::RawCalorimeterHit` (digi_hits), do you mean:
`sim_hits (input) > digitization -> digi_hits (intermediate) -> merging -> digi_hits (output)?`

File: src/algorithms/calorimetry/CalorimeterHitDigi.cc [hunk 1]
```diff
@@ -59,8 +59,8 @@ void CalorimeterHitDigi::init(const dd4hep::Detector* detector, std::shared_ptr<
     tRes       = m_cfg.tRes / dd4hep::ns;
     stepTDC    = dd4hep::ns / m_cfg.resolutionTDC;
 
+    id_mask = 0;
```
Comment (wdconinc) [inline src/algorithms/calorimetry/CalorimeterHitDigi.cc:62]: Can we at this point call this inverse mask with a docstring as to what the set bits are?

File: src/algorithms/calorimetry/CalorimeterHitDigi.cc [hunk 2]
```diff
@@ -86,64 +85,13 @@ void CalorimeterHitDigi::init(const dd4hep::Detector* detector, std::shared_ptr<
             // throw::runtime_error(fmt::format("Failed to load ID decoder for {}", m_cfg.readout));
             return;
         }
-        id_mask = ~id_mask;
         m_log->info("ID mask in {:s}: {:#064b}", m_cfg.readout, id_mask);
-        // all checks passed
-        merge_hits = true;
     }
+    id_mask = ~id_mask;
```
Comment (wdconinc) [inline src/algorithms/calorimetry/CalorimeterHitDigi.cc:90]: And then here it actually becomes the mask from the inverted mask.

**Review by wdconinc — APPROVED** [review]
This looks great from a code simplification point of view. Let's keep it in that scope.

It's a bit hard to interpret the diffs due to reordering. Can you copy-paste some diffs on the reconstructed clusters into the comments? As far as I can tell there shouldn't be any, right?

File: src/algorithms/calorimetry/CalorimeterHitDigi.cc (hunk 1 above)
Comment (veprbl) [reply src/algorithms/calorimetry/CalorimeterHitDigi.cc]: Fixed

**Review by wdconinc — APPROVED** [review]

**Comment by veprbl** [comment]
Let me comment on how this was tested. I apply a following change on top of this PR preserve the ordering:
```diff
diff --git a/src/algorithms/calorimetry/CalorimeterHitDigi.cc b/src/algorithms/calorimetry/CalorimeterHitDigi.cc
index 74e30f92..924b8432 100644
--- a/src/algorithms/calorimetry/CalorimeterHitDigi.cc
+++ b/src/algorithms/calorimetry/CalorimeterHitDigi.cc
@@ -97,6 +97,7 @@ std::unique_ptr<edm4hep::RawCalorimeterHitCollection> CalorimeterHitDigi::proces
     // find the hits that belong to the same group (for merging)
     std::unordered_map<uint64_t, std::vector<std::size_t>> merge_map;
     std::size_t ix = 0;
+    std::vector<uint64_t> hids;
     for (const auto &ahit : simhits) {
         uint64_t hid = ahit.getCellID() & id_mask;
 
@@ -104,13 +105,15 @@ std::unique_ptr<edm4hep::RawCalorimeterHitCollection> CalorimeterHitDigi::proces
         m_log->trace("new cell ID in {:s}: {:#064b}", m_cfg.readout, hid);
 
         merge_map[hid].push_back(ix);
+        hids.push_back(hid);
 
         ix++;
     }
 
     // signal sum
     // NOTE: we take the cellID of the most energetic hit in this group so it is a real cellID from an MC hit
-    for (const auto &[id, ixs] : merge_map) {
+    for (auto id : hids) {
+        auto ixs = merge_map[id];
         double edep     = 0;
         double time     = std::numeric_limits<double>::max();
         double max_edep = 0;
```
and this on top of PR base (main):
```diff
diff --git a/src/algorithms/calorimetry/CalorimeterHitDigi.cc b/src/algorithms/calorimetry/CalorimeterHitDigi.cc
index 6fe18466..eff6b9c9 100644
--- a/src/algorithms/calorimetry/CalorimeterHitDigi.cc
+++ b/src/algorithms/calorimetry/CalorimeterHitDigi.cc
@@ -110,6 +110,14 @@ std::unique_ptr<edm4hep::RawCalorimeterHitCollection> CalorimeterHitDigi::single
         // Note: juggler internal unit of energy is dd4hep::GeV
         const double eDep    = ahit.getEnergy();
 
+        double time = std::numeric_limits<double>::max();
+        for (const auto& c : ahit.getContributions()) {
+            if (c.getTime() <= time) {
+                time = c.getTime();
+            }
+        }
+        if (time > m_cfg.capTime) continue;
+
         // apply additional calorimeter noise to corrected energy deposit
         const double eResRel = (eDep > m_cfg.threshold)
                                ? m_normDist(generator) * std::sqrt(
@@ -118,18 +126,8 @@ std::unique_ptr<edm4hep::RawCalorimeterHitCollection> CalorimeterHitDigi::single
                                     std::pow(m_cfg.eRes[2] / (eDep), 2)
                                  )
                                : 0;
-
         const double ped    = m_cfg.pedMeanADC + m_normDist(generator) * m_cfg.pedSigmaADC;
         const long long adc = std::llround(ped + eDep * (m_cfg.corrMeanScale + eResRel) / m_cfg.dyRangeADC * m_cfg.capADC);
-
-        double time = std::numeric_limits<double>::max();
-        for (const auto& c : ahit.getContributions()) {
-            if (c.getTime() <= time) {
-                time = c.getTime();
-            }
-        }
-        if (time > m_cfg.capTime) continue;
-
         const long long tdc = std::llround((time + m_normDist(generator) * tRes) * stepTDC);
 
         if (eDep> 1.e-3) m_log->trace("E sim {} \t adc: {} \t time: {}\t maxtime: {} \t tdc: {} \t cell ID {}", eDep, adc, time, m_cfg.capTime, tdc, ahit.getCellID());
```
Then the only discrepancies are seen by capybara in ScFi and LFHCAL (that are, by construction, broken by the first test patch in this comment). This way, I conclude that this change does not affect the calorimeters that relied on `single_hits_digi` that is this PR removes.

**Pull request merged**